port/win/env_win.cc

```diff
@@ -0,0 +1,2604 @@
+      hFile_(hFile) {
+
+    assert(hFile != NULL);
+    assert(hFile != INVALID_HANDLE_VALUE);
+  }
+
+  ~WinFileLock()  {
+
+    BOOL ret = ::CloseHandle(hFile_);
+    assert(ret);
+  }
+
+private:
+  HANDLE hFile_;
+};
+
+namespace 
+{
+
+void WinthreadCall(const char* label, std::error_code result) {
+
+  if (0 != result.value()) {
+    fprintf(stderr, "pthread %s: %s\n", label, strerror(result.value()));
+    abort();
+  }
+}
+
+}
+
+class WinEnv : public Env {
+public:
+  WinEnv();
+
+  virtual ~WinEnv() {
+
+    for (auto& th : threads_to_join_) {
+      th.join();
+    }
+
+    threads_to_join_.clear();
+
+    for (auto& thpool : thread_pools_) {
+      thpool.JoinAllThreads();
+    }
+    // All threads must be joined before the deletion of
+    // thread_status_updater_.
+    delete thread_status_updater_;
+  }
+
+  virtual Status DeleteFile(const std::string& fname) override {
+
+    Status result;
+
+    if (_unlink(fname.c_str())) {
+      result = IOError("Failed to delete: " + fname, errno);
+    }
+
+    return result;
+  }
+
+  Status GetCurrentTime(int64_t* unix_time) override {
+
+    time_t time = std::time(nullptr);
+    if (time == (time_t)(-1)) {
+      return Status::NotSupported("Failed to get time");
+    }
+
+    *unix_time = time;
+    return Status::OK();
+  }
+
+  virtual Status NewSequentialFile(const std::string& fname, std::unique_ptr<SequentialFile>* result, const EnvOptions& options) override {
+
+    Status s;
+
+    result->reset();
+
+    // Corruption test needs to rename and delete files of these kind
+    // while they are still open with another handle. For that reason we
+    // allow share_write and delete(allows rename).
+    HANDLE hFile = 0;
+    {
+      IOSTATS_TIMER_GUARD(open_nanos);
+      hFile = CreateFileA(fname.c_str(),
+        GENERIC_READ,
+        FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+        NULL,
+        OPEN_EXISTING,         // Original fopen mode is "rb"
+        FILE_ATTRIBUTE_NORMAL,
+        NULL);
+    }
+
+    if (hFile == INVALID_HANDLE_VALUE) {
+      auto lastError = GetLastError();
+      s = IOErrorFromWindowsError("Failed to open NewSequentialFile" + fname, lastError);
+    } else {
+
+      int fd = _open_osfhandle(reinterpret_cast<intptr_t>(hFile), 0);
+      if (fd == -1) {
+        auto code = errno;
+        CloseHandle(hFile);
+        s = IOError("Failed to _open_osfhandle for NewSequentialFile: " + fname, code);
+      } else {
+
+        FILE * file = _fdopen(fd, "rb");
+        if (file == nullptr) {
+          auto code = errno;
+          _close(fd);
+          s = IOError("Failed to fdopen NewSequentialFile: " + fname, code);
+        } else {
+          result->reset(new WinSequentialFile(fname, file, options));
+        }
+      }
+    }
+    return s;
+  }
+
+  virtual Status NewRandomAccessFile(const std::string& fname, std::unique_ptr<RandomAccessFile>* result, const EnvOptions& options) override {
```

Comment:
@siying Cygwin bash refuses to run format-diff for some bizarre reason. 

$ build_tools/format-diff.sh
build_tools/format-diff.sh: line 108: syntax error: unexpected end of file

Would you be fine if I applied this patch on top on my most recent internal code review changes? I am asking because the patch includes some posix only files and others we did not change and they are not related to this port.